In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#Python shortcut for importing files, create string & assign to variable 
path = r'/Users/Chloe/Documents/2. CareerFoundry/2. Data Immersion pt 1/Achievement 4/07-2022 Instacart Basket Analysis'

In [4]:
path

'/Users/Chloe/Documents/2. CareerFoundry/2. Data Immersion pt 1/Achievement 4/07-2022 Instacart Basket Analysis'

In [5]:
#import the latest merged file using the os.path.join() function
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_v4.pkl'))

In [6]:
#check to be sure it includes the newest added columns from 4.7
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ... aisle_id  department_id  prices  _merge    price_range_loc  \
0          0  ...       77              7     9.0    both  Mid-range product   
1          1  ...       77              7     9.0    both  Mid-range product   
2          1  ...       77              7     9.0    both  Mid-range product   
3          1  ...       77              7     9.0    both  Mid-range product   
4          1  ...       77              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy      Least busy        Average orders        10   
2  Regularly busy      Least busy           Most orders        10   
3      Least busy      Least busy        Average orders        10   
4      Least busy      Least busy           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [7]:
ords_prods_merge.shape

(32404859, 21)

In [8]:
# Find the mean number of orders by department id (on the whole df)
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

We have the mean numbers for order numbers for all department IDs, not just for a subset of 1 mil rows.
We can note that there is a larger range between the department with the lowest avg (dept 5 with avg 15.2)
of order numbers and the department with the highest avg of order numbers (dept 21 with avg 22.9)

In [9]:
#Count frequency of customers by loyalty flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [10]:
#Check basis statistics of product prices for each loyalty category
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [11]:
#Look at individual customers based on spending to determine is whether the prices of products 
#purchased by loyal customers differ from those purchased by regular or new customers
# First: Create column of mean price by user
ords_prods_merge['mean_order'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [12]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0    2539329        1             1                   2                    8   
1    2398795        1             2                   3                    7   
2     473747        1             3                   3                   12   
3    2254736        1             4                   4                    7   
4     431534        1             5                   4                   15   
5    3367565        1             6                   2                    7   
6     550135        1             7                   1                    9   
7    3108588        1             8                   1                   14   
8    2295261        1             9                   1                   16   
9    2550362        1            10                   4                    8   
10   2968173       15            15                   1                    9   
11   1870022       15            17                   2                   16   
12   1911383       15            18                   2                   11   
13   2715276       15            21                   1                    9   
14    487368       15            22                   1                   10   

    days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                      NaN          True         196                  1   
1                     15.0         False         196                  1   
2                     21.0         False         196                  1   
3                     29.0         False         196                  1   
4                     28.0         False         196                  1   
5                     19.0         False         196                  1   
6                     20.0         False         196                  1   
7                     14.0         False         196                  2   
8                      0.0         False         196                  4   
9                     30.0         False         196                  1   
10                     7.0         False         196                  2   
11                     8.0         False         196                  6   
12                     7.0         False         196                  1   
13                     7.0         False         196                  1   
14                    14.0         False         196                  1   

    reordered  ... department_id  prices  _merge    price_range_loc  \
0           0  ...             7     9.0    both  Mid-range product   
1           1  ...             7     9.0    both  Mid-range product   
2           1  ...             7     9.0    both  Mid-range product   
3           1  ...             7     9.0    both  Mid-range product   
4           1  ...             7     9.0    both  Mid-range product   
5           1  ...             7     9.0    both  Mid-range product   
6           1  ...             7     9.0    both  Mid-range product   
7           1  ...             7     9.0    both  Mid-range product   
8           1  ...             7     9.0    both  Mid-range product   
9           1  ...             7     9.0    both  Mid-range product   
10          0  ...             7     9.0    both  Mid-range product   
11          1  ...             7     9.0    both  Mid-range product   
12          1  ...             7     9.0    both  Mid-range product   
13          1  ...             7     9.0    both  Mid-range product   
14          1  ...             7     9.0    both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy        Average orders        10   
1   Regularly busy      Least busy        Average orders        10   
2   Regularly busy      Least busy           Most orders        10   
3       Least busy      Least busy        Average orders        10   
4       Least busy      Least busy 

In [13]:
ords_prods_merge.shape

(32404859, 22)

In [14]:
# 2: Assign spending flag based on average price

ords_prods_merge.loc[ords_prods_merge['mean_order']<10, 'spending_flag'] = 'Low spender'

In [15]:
ords_prods_merge.loc[ords_prods_merge['mean_order']>=10, 'spending_flag'] = 'High spender'

In [16]:
ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [17]:
ords_prods_merge[['user_id', 'mean_order', 'spending_flag']].head(50)

user_id  mean_order spending_flag
0         1    6.367797   Low spender
1         1    6.367797   Low spender
2         1    6.367797   Low spender
3         1    6.367797   Low spender
4         1    6.367797   Low spender
5         1    6.367797   Low spender
6         1    6.367797   Low spender
7         1    6.367797   Low spender
8         1    6.367797   Low spender
9         1    6.367797   Low spender
10       15    3.980556   Low spender
11       15    3.980556   Low spender
12       15    3.980556   Low spender
13       15    3.980556   Low spender
14       15    3.980556   Low spender
15       19    8.372059   Low spender
16       19    8.372059   Low spender
17       19    8.372059   Low spender
18       21    7.012683   Low spender
19       31    8.137793   Low spender
20       31    8.137793   Low spender
21       43    7.444898   Low spender
22       43    7.444898   Low spender
23       52    6.356287   Low spender
24       52    6.356287   Low spender
25       52    6.356287   Low spender
26       52    6.356287   Low spender
27       52    6.356287   Low spender
28       52    6.356287   Low spender
29       52    6.356287   Low spender
30       52    6.356287   Low spender
31       52    6.356287   Low spender
32       52    6.356287   Low spender
33       52    6.356287   Low spender
34       52    6.356287   Low spender
35       52    6.356287   Low spender
36       52    6.356287   Low spender
37       67    7.656790   Low spender
38       67    7.656790   Low spender
39       67    7.656790   Low spender
40       67    7.656790   Low spender
41       67    7.656790   Low spender
42       67    7.656790   Low spender
43       67    7.656790   Low spender
44       67    7.656790   Low spender
45       67    7.656790   Low spender
46       67    7.656790   Low spender
47       67    7.656790   Low spender
48       67    7.656790   Low spender
49       67    7.656790   Low spender

In [18]:
# Create order frequency flag based on median days since previous order
#1: Create a new column showing median days since previous order
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [19]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ... _merge    price_range_loc     busiest_day    busiest_days  \
0          0  ...   both  Mid-range product  Regularly busy  Regularly busy   
1          1  ...   both  Mid-range product  Regularly busy      Least busy   
2          1  ...   both  Mid-range product  Regularly busy      Least busy   
3          1  ...   both  Mid-range product      Least busy      Least busy   
4          1  ...   both  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  loyalty_flag mean_order spending_flag  \
0        Average orders        10  New customer   6.367797   Low spender   
1        Average orders        10  New customer   6.367797   Low spender   
2           Most orders        10  New customer   6.367797   Low spender   
3        Average orders        10  New customer   6.367797   Low spender   
4           Most orders        10  New customer   6.367797   Low spender   

   median_days_since_prior_order  
0                           20.5  
1                           20.5  
2                           20.5  
3                           20.5  
4                           20.5  

[5 rows x 24 columns]

In [20]:
# Step two: Assign frequency flag based on median days since previous order
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order']>20, 'Frequency_flag'] = 'Non-frequent customer'

In [21]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order']>10) & (ords_prods_merge['median_days_since_prior_order']<=20), 'Frequency_flag'] = 'Regular customer'

In [22]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order']<=10, 'Frequency_flag'] = 'Frequent customer'

In [23]:
#check counts for new customer frequency flag created
ords_prods_merge['Frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: Frequency_flag, dtype: int64

In [24]:
ords_prods_merge[['user_id', 'median_days_since_prior_order', 'Frequency_flag']].head(100)

user_id  median_days_since_prior_order         Frequency_flag
0         1                           20.5  Non-frequent customer
1         1                           20.5  Non-frequent customer
2         1                           20.5  Non-frequent customer
3         1                           20.5  Non-frequent customer
4         1                           20.5  Non-frequent customer
..      ...                            ...                    ...
95      360                            4.0      Frequent customer
96      377                           16.5       Regular customer
97      387                            8.0      Frequent customer
98      420                            7.0      Frequent customer
99      420                            7.0      Frequent customer

[100 rows x 3 columns]

In [25]:
# Rename mean_order column to mean_order_price to make it easier to understand
ords_prods_merge.rename(columns = {'mean_order':'mean_order_price'}, inplace = True)

In [26]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy   
1          1  ...  Mid-range product  Regularly busy      Least busy   
2          1  ...  Mid-range product  Regularly busy      Least busy   
3          1  ...  Mid-range product      Least busy      Least busy   
4          1  ...  Mid-range product      Least busy      Least busy   

   busiest_period_of_day max_order  loyalty_flag mean_order_price  \
0         Average orders        10  New customer         6.367797   
1         Average orders        10  New customer         6.367797   
2            Most orders        10  New customer         6.367797   
3         Average orders        10  New customer         6.367797   
4            Most orders        10  New customer         6.367797   

  spending_flag median_days_since_prior_order         Frequency_flag  
0   Low spender                          20.5  Non-frequent customer  
1   Low spender                          20.5  Non-frequent customer  
2   Low spender                          20.5  Non-frequent customer  
3   Low spender                          20.5  Non-frequent customer  
4   Low spender                          20.5  Non-frequent customer  

[5 rows x 25 columns]

In [27]:
# adjust capitalization of Frequency_flag to frequency_flag for consistency
ords_prods_merge.rename(columns = {'Frequency_flag':'frequency_flag'}, inplace = True)

In [28]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy   
1          1  ...  Mid-range product  Regularly busy      Least busy   
2          1  ...  Mid-range product  Regularly busy      Least busy   
3          1  ...  Mid-range product      Least busy      Least busy   
4          1  ...  Mid-range product      Least busy      Least busy   

   busiest_period_of_day max_order  loyalty_flag mean_order_price  \
0         Average orders        10  New customer         6.367797   
1         Average orders        10  New customer         6.367797   
2            Most orders        10  New customer         6.367797   
3         Average orders        10  New customer         6.367797   
4            Most orders        10  New customer         6.367797   

  spending_flag median_days_since_prior_order         frequency_flag  
0   Low spender                          20.5  Non-frequent customer  
1   Low spender                          20.5  Non-frequent customer  
2   Low spender                          20.5  Non-frequent customer  
3   Low spender                          20.5  Non-frequent customer  
4   Low spender                          20.5  Non-frequent customer  

[5 rows x 25 columns]

In [29]:
# Export newest file with updates as pickle
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_v5.pkl'))